libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +630,246 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+{
+	for (ModifierDefinition const* modifier: _modifiers)
+	{
+		if (contains_if(_funcSet, MatchByName{modifier->name()}))
+			m_errorReporter.typeError(
+				modifier->location(),
+				"Override changes function to modifier."
+			);
+
+		auto [begin,end] = _modSet.equal_range(modifier);
+
+		// Skip if no modifiers found in bases
+		if (begin == end)
+			continue;
+
+		if (!modifier->overrides())
+			overrideError(*modifier, **begin, "Overriding modifier is missing 'override' specifier.");
+
+		for (; begin != end; begin++)
+			if (ModifierType(**begin) != ModifierType(*modifier))
+				m_errorReporter.typeError(
+					modifier->location(),
+					"Override changes modifier signature."
+			);
+	}
+
+}
+
+void ContractLevelChecker::checkOverrideList(FunctionSet const& _funcSet, FunctionDefinition const& _function)
+{
+	set<ContractDefinition const*> specifiedContracts =
+		_function.overrides() ?
+		resolveOverrideList(*_function.overrides()) :
+		set<ContractDefinition const*>{};
+
+	// Check for duplicates in override list
+	if (_function.overrides() && specifiedContracts.size() != _function.overrides()->overrides().size())
+	{
+		// Sort by name to find duplicate for error reporting
+		vector<ASTPointer<UserDefinedTypeName>> list =
+			sortByName(_function.overrides()->overrides());
+
+		// Find duplicates and output error
+		for (size_t i = 1; i < list.size(); i++)
+			if (list[i]->namePath() == list[i-1]->namePath())
+			{
+				SecondarySourceLocation ssl;
+				ssl.append("First occurrence here: ", list[i-1]->location());
+				m_errorReporter.typeError(
+					list[i]->location(),
+					ssl,
+					"Duplicate contract " +
+						joinHumanReadable(list[i]->namePath(), ".") +
+						" found in override list of " +
+						_function.name() +
+						"."
+				);
+			}
+	}
+
+	decltype(specifiedContracts) missingContracts;
+
+	// Iterate over the overrides
+	for (auto [begin, end] = _funcSet.equal_range(&_function); begin != end; begin++)
+	{
+		// Validate the override
+		if (!checkFunctionOverride(_function, **begin))
+			break;
+
+		auto const result = find(
+			specifiedContracts.cbegin(),
+			specifiedContracts.cend(),
+			(*begin)->annotation().contract
+		);
+
+		if (result == specifiedContracts.cend())
+			missingContracts.insert((*begin)->annotation().contract);
+		else
+			specifiedContracts.erase(result);
+	}
+
+	if (missingContracts.size() > 1)
+		overrideListError(
+			_function,
+			missingContracts,
+			"Function needs to specify overridden ",
+			""
+		);
+
+	if (!specifiedContracts.empty())
+		overrideListError(
+			_function,
+			specifiedContracts,
+			"Invalid ",
+			"specified in override list: "
+		);
+}
+
+ContractLevelChecker::FunctionSet const& ContractLevelChecker::getBaseFunctions(ContractDefinition const* _contract) const
+{
+	auto const& result = m_contractBaseFunctions.find(_contract);
```

Comment:
Suggestion to make it more obvious that this is a cache that is generated on the fly:
```
if (!m_contractBaseFunctions.count(_contract))
{
  // compute stuff and assign to `m_contractBaseFunctions[_contract]`.
}
return m_contractBaseFunctions[_contract];
```

Reply:
I reverted the `if` condition, but I don't like to use `count()` and `m_contractBaseFunctions[..]`. Count iterates through everything even though it can return at the first result and `[] searches it all over again even though we did just that a few lines up..

Reply:
If you do not like the added complexity of `count`, you can use some of the `contains` helper functions we have. Whether or not performing a second access has any impact at all on runtime is not known. In any case, readability should outweigh runtime complexity. If the compiler is slow on certain inputs, we should profile it and improve that places.